support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -1,12 +1,79 @@
 package com.gu.support.workers
 
+import cats.implicits._
 import com.gu.i18n.Title
 import com.gu.support.encoding.Codec
 import com.gu.support.encoding.Codec.deriveCodec
 import com.gu.support.encoding.CustomCodecs._
+import com.gu.support.workers.GiftRecipient.{DigiSub, Weekly}
+import io.circe._
+import io.circe.syntax._
 
-case class GiftRecipient(title: Option[Title], firstName: String, lastName: String, email: Option[String])
-
+sealed trait GiftRecipient {
+  val giftRecipientType: String
```

Comment:
this I was a bit up in the air, so this whole file is a bit of a FIXME.

We need to keep different info for weekly and DS giftees, so I introduced two subclasses here (rather than the everything optional approach as that just pushes the validation later in a "this can't happen" kind of way)

however the subclasses are compatible, which falls foul of the default "try one and then try the other" way.  i.e. if it has first+last+email, the deserialiser has no idea if it's a weekly or a DS giftee.

I decided to add a discrimintor to the JSON but I couldn't decide whether to use the standard semiauto and expose in the case classes, or whether to make it deterministic and select the deserialiser in a custom Codec. ( I have implemented the built in option, and roughly the deterministic option is roughly commented out below)

Reply:
rupert and I have discussed and decided that firstly the default fields thing doesn't actually do what I expected, the default value is ignored so it would just deserialise to the wrong one anyway.  We decided that the deterministic approach is better and it avoids polluting the scala model with serialisation specific stuff.